### PR TITLE
Update Software Engineering Upgradeability section

### PR DESCRIPTION
**Pull request by fodisi**
* Update Software Engineering Upgradeability section, addressing the issue #164 

Registry
* Update sample to Solidity 0.5.0
* Add address validation "changeBackend" function

Relay
* Add risks and recommendations regarding the use of delegatecall
* Addresses security issues regarding proxies
* Add LogicContract sample
* Update sample to Solidity 0.5.0
* Add address validation "constructor" and "payable" function

### CODE REVIEW

**Comment by maurelian** [comment]
Thanks @fodisi! One of us will review this when we get a chance next week.

File: docs/software_engineering.md [hunk 1]
```diff
@@ -71,26 +80,51 @@ contract Relay {
...
+- Why a new version of the called contract (`LogicContract`) [must have the same storage layout as the previous version](https://github.com/OpenZeppelin/openzeppelin-sdk/issues/37), and only append new variables to the storage
+- [How a contract's constructor can affect upgradability](https://blog.openzeppelin.com/towards-frictionless-upgradeability/)
+- How the ABI specifies [function selectors](https://solidity.readthedocs.io/en/v0.4.24/abi-spec.html?highlight=signature#function-selector) and how [function-name collision](https://medium.com/nomic-labs-blog/malicious-backdoors-in-ethereum-proxies-62629adf3357) can be used to exploit a contract that uses `delegatecall`
+- How `delegatecall` to a non-existent contract will return true even if the called contract does no exist. For more details see [Breaking the proxy pattern](https://blog.trailofbits.com/2018/09/05/contract-upgrade-anti-patterns/) and Solidity docs on [Error handling](https://solidity.readthedocs.io/en/latest/control-structures.html#error-handling-assert-require-revert-and-exceptions).
+
```
Comment (shayanb) [inline docs/software_engineering.md]: I suggest adding this blog post to the resources as well: https://diligence.consensys.net/blog/2019/01/upgradeability-is-a-bug/

Comment (fodisi) [reply docs/software_engineering.md]: @shayanb 
Thank you for your review. I just a reference to the blog post provided.

**Comment by fodisi** [comment]
@shayanb @maurelian 
Hey guys. Is there anything else you'd like me to handle/change related to this PR?
If there's anything that's preventing it from being merged, just let me know and I'll gladly rework it.

Thank you!

**Comment by shayanb** [comment]
@fodisi thanks for the PR and also reminding us to merge it :)

**Pull request merged**